lib/labels/labels.go

```diff
@@ -166,3 +169,118 @@ func (l *Dynamic) setLabel(name string, value types.CommandLabel) {
 
 	l.c.Labels[name] = value
 }
+
+// EC2LabelConfig is the configuration for the EC2 label service.
+type EC2LabelConfig struct {
+	Client utils.InstanceMetadata
+	Clock  clockwork.Clock
+	Log    *logrus.Entry
+}
+
+func (conf *EC2LabelConfig) checkAndSetDefaults() error {
+	if conf.Client == nil {
+		client, err := utils.NewInstanceMetadataClient(context.TODO())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		conf.Client = client
+	}
+	if conf.Clock == nil {
+		conf.Clock = clockwork.NewRealClock()
+	}
+	if conf.Log == nil {
+		conf.Log = logrus.NewEntry(logrus.StandardLogger())
+	}
+	return nil
+}
+
+// EC2Labels is a service that periodically imports tags from EC2 via instance
+// metadata.
+type EC2Labels struct {
+	c      *EC2LabelConfig
+	mu     sync.RWMutex
+	once   sync.Once
+	labels map[string]string
+
+	closeContext context.Context
+	closeFunc    context.CancelFunc
+}
+
+func NewEC2Labels(ctx context.Context, c *EC2LabelConfig) (*EC2Labels, error) {
+	if err := c.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	closeContext, closeFunc := context.WithCancel(ctx)
+
+	return &EC2Labels{
+		c:            c,
+		labels:       make(map[string]string),
+		closeContext: closeContext,
+		closeFunc:    closeFunc,
+	}, nil
+}
+
+// Get returns the list of updated EC2 labels.
+func (l *EC2Labels) Get() map[string]string {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+	return toAWSLabels(l.labels)
+}
+
+// Sync will block and synchronously update EC2 labels.
+func (l *EC2Labels) Sync() {
+	m := make(map[string]string)
+
+	tags, err := l.c.Client.GetTagKeys()
+	if err != nil {
+		l.c.Log.Errorf("Error fetching EC2 tags: %v", err)
+		return
+	}
+
+	for _, t := range tags {
+		value, err := l.c.Client.GetTagValue(t)
+		if err != nil {
+			l.c.Log.Errorf("Error fetching EC2 tags: %v", err)
+			return
+		}
+		m[t] = value
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	l.labels = m
+}
+
+// Start will start a loop that continually keeps EC2 labels updated. Start
+// can safely be called multiple times.
+func (l *EC2Labels) Start() {
+	l.once.Do(func() { go l.periodicUpdateLabels() })
+}
+
+func (l *EC2Labels) periodicUpdateLabels() {
+	ticker := l.c.Clock.NewTicker(types.EC2LabelUpdatePeriod)
+	defer ticker.Stop()
+
+	for {
+		l.Sync()
+		select {
+		case <-ticker.Chan():
+		case <-l.closeContext.Done():
+			return
+		}
+	}
+}
+
+// Close will free up all resources and stop keeping EC2 labels updated.
+func (l *EC2Labels) Close() {
+	l.closeFunc()
+}
+
+// toAWSLabels formats labels coming from EC2.
+func toAWSLabels(labels map[string]string) map[string]string {
+	m := make(map[string]string, len(labels))
+	for k, v := range labels {
+		m[fmt.Sprintf("%s/%s", types.AWSNamespace, k)] = v
```

Comment:
Could you just save them this way when you read the tags from the IMDS, rather than convert the labels every time they are read?